### DISCUSSION TITLE
Optimism's fast transaction throughput reasons

### DISCUSSION THREAD

**Discussion by novaknole** [ethereum-optimism/developers#304] [Core Protocol] [answered]
### Did you check the documentation?

- [X] Yes

### Did you read the specs?

- [X] Yes
- [ ] No

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

This question will be mostly related to optimism v0 version and the current one(where sequencer is still centralized). I'd appreciate so much if you could answer each question one by one, so it's easier for me to grasp your thoughts. Thanks a ton.

My main question is: How does optimism allow faster transaction throughput ? and I will explain what my basis to this question is.

So, on L2, there's a sequencer node and normal nodes. Sequencer node is the only one that can publish the block to L2's normal nodes and sequencer is the only one that can publish calldatas and state roots to L1 chain. Before moving forward, in my opinion, this is what makes things super fast whereas in ethereum(proof of work), since every node can solve the blocks and share it, ethereum came up with proof of work algorithm to reduce frequent forks and ensure double-spend problem doesn't occur.  This is why ethereum couldn't handle fast transaction throughput and not only that, if ethereum increased block size, this would result in less decentralization due to higher hard drive requirements.

As I understand, in optimism, sequencer publishes txs to other normal nodes where normal nodes can only run them in their own L2 EVM and apply those txs to their state db and other normal nodes can't be re-publishing  blocks or txs.  Because of this, there's no need for proof of work which results in no more frequent forks and faster transaction throughput on L2. 

**Question 1: Before I move forward, i want to ask: is it the correct assumption that optimism allows faster transaction throughput because it doesn't need to run proof of work consensus or proof of stake consensus where block time is some fixed value(i.e 12 sec) for example ?** 

**Question 2: Since L2 normal nodes listen to transactions and update their L2 chains and since L2 blocks contain transactions, sooner or later, it will be hard to join L2 chain as a normal node because of hard drive requirements as L2 blockchain increases as time goes. Won't this result in less decentralization ?** 

**Question 3: Once you make the sequencer decentralized, that means many people will be able to run it which means optimism will result in the same problems as ethereum did. This is because if there're many sequencers, they can share the block at the same time and all the problems of frequent forks, propagation time, block time and so on will happen for L2 chain as well.  Am I wrong ?**

**Comment by hash1go** [comment, marked as answer]
Hello!

> Question 1: Before I move forward, i want to ask: is it the correct assumption that optimism allows faster transaction throughput because it doesn't need to run proof of work consensus or proof of stake consensus where block time is some fixed value(i.e 12 sec) for example ?

It's important to clarify that the transaction throughput of OP Mainnet is not significantly higher compared to Ethereum. While the differences in the EIP-1559 parameters do provide a greater ability to cover temporary spikes in demand, in the long term, the processing capacity is only about twice that of Ethereum. I recommend referring to [this document](https://docs.optimism.io/chain/differences#eip-1559-parameters) or [my blog](https://mirror.xyz/hashigo%F0%9F%94%B4.eth/sZ42ZTmmEzC99gTQ7djtkec6AOXcoJNxgLO6wpHfDXA) for more detailed information.

Also, it's noteworthy that OP Mainnet has a fixed block time of 2 seconds, and the absence of a consensus mechanism does not directly translate to an increase in transaction throughput.


> Question 2: Since L2 normal nodes listen to transactions and update their L2 chains and since L2 blocks contain transactions, sooner or later, it will be hard to join L2 chain as a normal node because of hard drive requirements as L2 blockchain increases as time goes. Won't this result in less decentralization ?

Indeed, over time, the size of the state will increase, which is a common issue and not specific to OP Mainnet. However, the size is still within the manageable range of SSD capacities, and SSDs are more appropriate for nodes as hard drives are less efficient in terms of random access performance.

It's expected that as the requirements for running a node become more challenging, fewer people might engage in setting up nodes. However, it's important to note that OP Mainnet is a rollup, and the situation is different from PoS or PoW chains. This distinction is significant when considering the implications for decentralization.

> Question 3: Once you make the sequencer decentralized, that means many people will be able to run it which means optimism will result in the same problems as ethereum did. This is because if there're many sequencers, they can share the block at the same time and all the problems of frequent forks, propagation time, block time and so on will happen for L2 chain as well. Am I wrong ?

Decentralizing the sequencer is indeed part of the future roadmap, but specific details have not been made public yet. The answer to your question would depend on the method of achieving this decentralization.

**Reply by novaknole** [reply]
Amazing answers @hash1go. 

**Q1:** Since OP mainnet mentions that sequencer is the only block producer,  what do "normal nodes that can run on L2" actually do ? They can listen incoming blocks from sequencer, but even if they realize that sequencer's block is malicious, they can't do anything until block's txs calldata is submitted to L1 chain and then fraud proof begins. Long story short, what I am saying is 1 sequencer and 1 verifier is already enough currently. What do normal nodes actually do currently ?

**Q2:** If the block time is 2secs, I guess, as soon as sequencer creates a block, it also pushes its txs' calldatas to L1 chain. What was the reason OP chose 2 seconds ? I mean, for one thing, maybe you wanted other nodes to have time to receive blocks from sequencer, but if normal nodes don't produce blocks, there will NOT be forks, hence even if blocks are solved 1.5 sec, the situation would be the same.

**Reply by hash1go** [reply]
> Q1: Since OP mainnet mentions that sequencer is the only block producer, what do "normal nodes that can run on L2" actually do ? They can listen incoming blocks from sequencer, but even if they realize that sequencer's block is malicious, they can't do anything until block's txs calldata is submitted to L1 chain and then fraud proof begins. Long story short, what I am saying is 1 sequencer and 1 verifier is already enough currently. What do normal nodes actually do currently ?

Firstly, it's important to understand that the normal nodes in the OP mainnet are divided into two types: op-geth and op-node. The op-geth nodes are akin to the execution layer in Ethereum, whereas op-node functions similarly to Ethereum's consensus layer. The op-geth RPC is responsible for receiving transactions and submitting them to the sequencer. Blocks distributed by the sequencer via P2P networking are then received by op-node, which subsequently inserts them into op-geth. In this context, the normal nodes in the OP Mainnet can be interpreted as standard full nodes that maintain the world state.

> Q2: If the block time is 2secs, I guess, as soon as sequencer creates a block, it also pushes its txs' calldatas to L1 chain. What was the reason OP chose 2 seconds ? I mean, for one thing, maybe you wanted other nodes to have time to receive blocks from sequencer, but if normal nodes don't produce blocks, there will NOT be forks, hence even if blocks are solved 1.5 sec, the situation would be the same.

When a block is created, its calldata is not immediately sent to L1. Instead, compressed transaction batches are transmitted to L1 approximately every 5 minutes. You can refer to [this explorer](https://blobscan.com/address/0x6887246668a3b87f54deb3b94ba47a6f63f32985) for more details.

Regarding the choice of a 2secs block time in OP Mainnet, this is my personal opinion: I believe the decision was based on balancing the time required to broadcast blocks globally with the need for a quick confirmation UX. In this context, a block time of 2 seconds appears to be an optimal compromise. A very short block time could potentially cause discrepancies in block heights across nodes around the world.

**Reply by novaknole** [reply]
@hash1go 
1. It's clear that there're 2 node types. user's tx first goes to op-geth node which forwards it to sequencer, whereas produced L2 block is given to op-node which puts it in its chain as well as forwards it to op-geth. The tricky part is when op-node receives a L2 block from sequencer, does it start validating immediatelly ? I guess so and if so, what if it finds the malicious behaviour ? in this case, op-node won't add it to its chain and other op-nodes won't do it either. Even if that happens, sequencer can still submit it to L1 and then op-nodes or verifiers will start fraud proof. So basically, the responsibility i see with op-nodes is that they maintain the L2 ledger, so if user asks what his balance is, op-node is the one that gives it to him. I thought that single sequencer could handle all of this, but his rpc wouldn't handle this many requets. I think it makes sense. Thoughts ? have I said anything incorrect ? 

2.  Since there could be 30 op-nodes all over the world, is 2 seconds enough for the sequencer's block propagation to arrive to 30 op-nodes ? it could be that 20 op-nodes receive the block, while 10 op-nodes didn't receive it in that 2 seconds and what could happen is sequencer might publish and send the next block and somehow this block might arrive to those 10 op-nodes. What we will get is: 20 op-nodes will have A->B->C, whereas 10 op-nodes will have A only(these 10 op-nodes won't add C because its parent hash doesn't correspond to A). So we still got inconsistency. Does the longest chain rule still apply in L2 ? if so, we still got some kind of consensus for L2, doesn't it ?

**Reply by hash1go** [reply]
> 1.

Your understanding is correct. If a sequencer submits an invalid state root to L1, the full nodes on the OP Mainnet are capable of detecting this improper state transition. Because the OP Mainnet's trusted sources are present on L1, it becomes necessary to correct any fraudulent state roots through a fault proof process. (However, it's important to note that the fault proof system is currently available only on the testnet and is still under development.)

> 2.

My earlier explanation about the 2secs block time being an optimal compromise is just my opinion and could be incorrect. However, when we look at chains with short block times, like Polygon, I believe that having a short block time can trigger network instability. Moreover, in the L2, the longest chain rule is not implemented; instead, trusted sources are recorded on L1.

**Reply by novaknole** [reply]
@hash1go 
1. Yes, but 2 sec block time definitely might cause instability on L2 as well. propagation might take more than 2 secs definitely which will also result instability on OP Mainnet as well.  I guess, since longest chain rule doesn't exist, the difference in L2's node chains are resolved by fetching submitted txs from L1. 

2. So, on Optimism, longest chain rule, proof of work/proof of stake doesn't exist. What else doesn't exist compared to ethereum ? I'd appreciate the list of things that OP doesn't take from ethereum. Do L2 blocks store state root as well ? Does block validation look the same in L2 ? and so on.

**Reply by hash1go** [reply]
It's difficult to list all the differences between Ethereum and OP Mainnet. For example, you can refer to [this document](https://docs.optimism.io/chain/differences).

**Reply by novaknole** [reply]
It doesn't list anything related to blocks. Does this mean that L2 blocks contain the exact same thing as L1 ? (mainly, I am curious about merkle roots), but no worries if you don't know.

**Reply by hash1go** [reply]
L2 block contains the merkle roots.
https://specs.optimism.io/glossary.html#block

**Reply by novaknole** [reply]
@hash1go 

I was just thinking, whether if optimism chose 0.5 sec block time instead of 2 sec,  would this cause forks between L2 nodes ? Since there's only one block producer, I don't think so. Do you ? In my opinion, What could happen is some nodes would receive blockA, some nodes wouldn't, but blockA would still reach those. I don't think inconsistency would continue more than 5-10 seconds and that would be inconsistent only 1 block height.

**Reply by hash1go** [reply]
Forks would not occur

**Reply by novaknole** [reply]
So then, I still don't understand why 2 sec block time and not 0.5 sec. do you have an opinion on this ? This is what you said previously:

> I believe the decision was based on balancing the time required to broadcast blocks globally with the need for a quick confirmation UX. In this context, a block time of 2 seconds appears to be an optimal compromise.

Even if blocks are not broadcasted globally(i.e propagation might take more time than block producing happens - not always, but sometimes), then what would go wrong with UI ? i don't get it.

**Reply by hash1go** [reply]
Indeed, a shorter block time leads to faster confirmations, which can improve the user experience. However, it's important to consider the increased load on network nodes. Shorter block times can result in more frequent block generation, increasing the computational and communication burden on the nodes.

**Reply by novaknole** [reply]
@hash1go  thanks so much for all the help. You rock, man !!! I will give some thoughts about all the things we talked about and let's see where it goes.

I somehow have confusing questions sometimes, it seems easy at first glance, but when I think about it in a deep way, I end up in a confusing state. I was also curious about state roots which I asked here - https://github.com/ethereum-optimism/developers/discussions/305 no rush, but it's an interesting question in my opinion.

**Reply by hash1go** [reply]
You're welcome. If there are no further questions on this topic, please select the post that answered your question to close this discussion.

**Reply by sbvegan** [reply]
Thank you @hash1go for your support here. @novaknole in regard to the 2 second block time because op-geth has a [minimal diff](https://op-geth.optimism.io/) with geth, the block time needs to be a whole number. This is just based on the how the code handles block times at the moment. So technically we could drop it to 1s or increase the diff to allow for even faster block times. Just a little additional insight.
